test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
+      });
 
-      sessionPool.reads_ = fakeReads;
-      sessionPool.writes_ = fakeWrites;
-      sessionPool.borrowed_ = fakeBorrowed;
+      it('should get a read session', function() {
+        sessionPool.createTransaction_ = () => Promise.resolve();
+        sessionPool.session_ = function() {
+          return new Object({
+            lastUsed: Date.now(),
+            create: () => Promise.resolve(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        sessionPool.open();
+        return new Promise(resolve => setTimeout(resolve, 300)).then(() =>
+          sessionPool.getReadSession().then(function(s) {
+            assert.strictEqual(s.type, 'readonly');
+            sessionPool.release(s);
+          })
+        );
+      });
 
-      var destroyCallCount = 0;
-      sessionPool.destroySession_ = function(session) {
-        assert.strictEqual(session, fakeAll[destroyCallCount++]);
-      };
+      it('should get a write session', function() {
+        sessionPool.createTransaction_ = () => Promise.resolve();
+        sessionPool.session_ = function() {
+          return new Object({
+            create: () => Promise.resolve(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        sessionPool.open();
+        return new Promise(resolve => setTimeout(resolve, 300)).then(() =>
+          sessionPool.getWriteSession().then(function(s) {
+            assert.strictEqual(s.type, 'readwrite');
+            sessionPool.release(s);
+          })
+        );
+      });
 
-      sessionPool.close();
+      it('should not get a read session as database is closed', function() {
+        sessionPool.session_ = function() {
+          return new Object({
+            create: () => Promise.reject(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        return sessionPool.getReadSession().catch(error => {
+          assert.strictEqual(error.message, 'Database is closed.');
+        });
+      });
 
-      assert.strictEqual(destroyCallCount, fakeAll.length);
-      assert.deepEqual(sessionPool.reads_, []);
-      assert.deepEqual(sessionPool.writes_, []);
-      assert.deepEqual(sessionPool.borrowed_, []);
+      it('should not get a write session as database is closed', function() {
+        sessionPool.session_ = function() {
+          return new Object({
+            create: () => Promise.reject(),
+            delete: () => Promise.resolve(),
+          });
+        };
+        return sessionPool.getWriteSession().catch(error => {
+          assert.strictEqual(error.message, 'Database is closed.');
+        });
+      });
     });
 
-    it('should settle once all sessions are destroyed', function() {
-      var delay = 500;
-
-      sessionPool.reads_ = [{}];
-      sessionPool.destroySession_ = function() {
-        return new Promise(function(resolve) {
-          setTimeout(resolve, delay);
+    describe('release', function() {
```

Comment:
Can we make a test for each type of resource (read, read/write) that ensures release is called against the correct underlying pool?

Reply:
updated the test case